kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/KafkaService.java

```diff
@@ -720,4 +723,133 @@ private InternalClusterMetrics calculateClusterMetrics(
+      for (var assignmentList : currentAssignment.values()) {
+        sortedBrokers = getMapSortedByValues(sortedBrokers);
+
+        // Iterate brokers and try to add them in assignment
+        // while (partition replicas count != requested replication factor)
+        for (Map.Entry<Integer, Integer> broker : sortedBrokers.entrySet()) {
+          if (!assignmentList.contains(broker.getKey())) {
+            assignmentList.add(broker.getKey());
+            sortedBrokers.put(broker.getKey(), broker.getValue() + 1);
+          }
+          if (assignmentList.size() == replicationFactorChange.getTotalReplicationFactor()) {
+            break;
+          }
+        }
+        if (assignmentList.size() != replicationFactorChange.getTotalReplicationFactor()) {
+          throw new ValidationException("Something went wrong while adding replicas");
+        }
+      }
+
+      // If we should to decrease Replication factor
+    } else if (replicationFactorChange.getTotalReplicationFactor() < currentReplicationFactor) {
+
+      // For each partition remove replicas from the end of current assignment
+      // while (partition replicas count != requested replication factor)
+      currentAssignment.values().forEach(assignmentList -> {
+        while (assignmentList.size() != replicationFactorChange.getTotalReplicationFactor()) {
+          assignmentList.remove(assignmentList.size() - 1);
+        }
+      });
+    } else {
+      throw new ValidationException("Replication factor already equals requested");
+    }
+
+    // Fill result map
+    Map<TopicPartition, Optional<NewPartitionReassignment>> reassignments = new HashMap<>();
+    currentAssignment.forEach((key, value) -> reassignments.put(
+        new TopicPartition(topicName, key),
+        Optional.of(new NewPartitionReassignment(value))));
+    return reassignments;
+  }
+
+  private Map<Integer, List<Integer>> getCurrentAssignment(KafkaCluster cluster, String topicName) {
+    Map<Integer, List<Integer>> currentAssignment = new HashMap<>();
+
+    cluster.getTopics().get(topicName).getPartitions().values()
+        .forEach(partition -> currentAssignment.put(partition.getPartition(),
+            partition.getReplicas().stream()
+                .map(InternalReplica::getBroker)
+                .collect(Collectors.toList())));
+
+    return currentAssignment;
+  }
+
+  private Map<Integer, Integer> getBrokersMap(KafkaCluster cluster, String topicName) {
+    Map<Integer, Integer> brokers = new LinkedHashMap<>();
+
+    cluster.getMetrics().getInternalBrokerDiskUsage().keySet()
+        .forEach(id -> brokers.put(id, 0));
+
+    cluster.getTopics().get(topicName).getPartitions().values().stream()
+        .flatMap(t -> t.getReplicas().stream())
+        .map(InternalReplica::getBroker)
+        .forEach(broker -> brokers.put(broker, brokers.get(broker) + 1));
+
+    return brokers;
+  }
+
+  private Map<Integer, Integer> getMapSortedByValues(Map<Integer, Integer> map) {
```

Comment:
Maybe better move util  functions to Util classes

Reply:
This function deleted because it's not actual